wafer/kvpairs/models.py

```diff
@@ -0,0 +1,252 @@
+                self.model_ct.model)
+    # For direct use of this attribute in forms:
+    get_model_name.short_description = model_ct.verbose_name
+    get_model_name.admin_order_field = 'model_ct'
+
+    def get_fully_qualified_key_name(self, quoted=False):
+        '''Return the name of the key properly namespaced'''
+        name = self.name.join(("'",)*2) if quoted else self.name
+        return '/'.join((self.get_model_name(), name))
+    # For direct use of this attribute in forms:
+    get_fully_qualified_key_name.short_description = name.verbose_name
+    get_fully_qualified_key_name.admin_order_field = 'name'
+
+    def is_compatible_instance(self, instance):
+        '''Return whether a given instance is compatible with the associated model'''
+        return isinstance(instance, self.get_model_class())
+
+    def natural_key(self):
+        '''Return the natural key of this Key (for serialization)'''
+        return self.model_ct.natural_key() + (self.name,)
+
+    def __str__(self):
+        '''Return an informative text display string for this Key'''
+        return u"[%s] %s" % (
+                'id:%s'%self.id if self.id else 'unsaved',
+                self.get_fully_qualified_key_name(quoted=True))
+
+
+@python_2_unicode_compatible
+class KeyValuePair(models.Model):
+    """Mapping of referenced model instances to values"""
+
+    # The key definition to use
+    key = models.ForeignKey('Key', on_delete=models.PROTECT,
+            verbose_name='Base key context')
+
+    # The foreign key to the actual object we're referencing, e.g. the User to
+    # which we are attributing a value
+    ref_id = models.PositiveIntegerField(
+            verbose_name='ID of referenced model instance')
+
+    # The logic for the indirect generic foreign key happens in the
+    # IndirectGenericForeignKey field, which we need to instantiate, but it
+    # doesn't actually hold any data.
+    ref_obj = IndirectGenericForeignKey(ct_field_path='key.model_ct',
+            fk_field='ref_id')
+
+    # finally the value to store
+    value = models.CharField(max_length=VALUE_MAXLEN,
+            verbose_name='Value for key associated with model instance')
+
+    class Meta:
+        unique_together = (('key', 'ref_id'),)
+        verbose_name = 'Key-value pair'
+
+    def __str__(self):
+        '''Return an informative text display string for this KeyValuePair'''
+        return u'[%s] %s(%s)/\'%s\'="%s"' % (
+                'id:%s'%self.id if self.id else 'unsaved',
+                self.key.get_model_name(), self.get_ref_instance(),
+                self.key.name, self.value)
+
+    def get_model_class(self):
+        '''Returns the model class associated with the Key'''
+        return self.key.get_model_class()
+
+    def get_ref_instance(self):
+        '''Returns the instance referenced by the KeyValuePair'''
+        model_class = self.get_model_class()
+        return model_class.objects.get(pk=self.ref_id)
```

Comment:
This looks a lookup that could be cached - we don't expect the referenced object to change and this seems like it will be called reasonably often if working with the value pairs.